### PR TITLE
Add test for last written lsn cache

### PR DESCRIPTION
**Pull request by knizhnik**
Main:
```
test_measure_read_latency_heavy_write_workload[neon-1].run_duration: 29.455 s
test_measure_read_latency_heavy_write_workload[neon-1].read_latency_avg: 4.346 s
test_measure_read_latency_heavy_write_workload[neon-1].read_latency_stdev: 5.499 s
```

last_written_lsn_cache:
```
test_measure_read_latency_heavy_write_workload[neon-1].run_duration: 34.074 s
test_measure_read_latency_heavy_write_workload[neon-1].read_latency_avg: 0.422 s
test_measure_read_latency_heavy_write_workload[neon-1].read_latency_stdev: 0.761 s
```

### CODE REVIEW

**Comment by aome510** [comment]
Look like this branch is based on the code in #1891 as `record_read_latency` isn't defined anywhere in `main`.

The new test shares some similarities with the tests added in #1891. I think we can use tests in #1891 to test the `last_written_lsn_cache` implementation.

Also, looks like CircleCI doesn't run because "We have detected an anomaly that violates the [CircleCI Terms of Service](https://support.circleci.com/hc/en-us/articles/360043679453-CircleCI-Terms-of-Service-Violation-Sanctioned-Country). [Troubleshoot](https://support.circleci.com/hc/en-us/sections/115003756627-Troubleshooting-and-Common-Issues)", which is probably because the PR is submitted from Russia.

File: test_runner/performance/test_latency.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+from fixtures.neon_fixtures import Postgres
+from performance.test_perf_pgbench import get_scales_matrix
+from performance.test_wal_backpressure import record_read_latency
+
+
+def start_write_workload(pg: Postgres, scale: int = 10):
+    with pg.connect().cursor() as cur:
+        cur.execute(f"create table big as select generate_series(1,{scale*100_000})")
+
+
```
Comment (hlinnaka) [inline test_runner/performance/test_latency.py:14]:
```suggestion
# Measure latency of reads on one table, while lots of writes are happening on another table.
# The fine-grained tracking of last-written LSNs helps to keep the latency low. Without it, the reads would
# often need to wait for the WAL records of the unrelated writes to be processed by the pageserver.
```

File: test_runner/regress/test_lsn_mapping.py [hunk 2]
```diff
@@ -34,9 +34,11 @@ def test_lsn_mapping(neon_env_builder: NeonEnvBuilder):
 
     # Execute one more transaction with synchronous_commit enabled, to flush
     # all the previous transactions
-    cur.execute("SET synchronous_commit=on")
     cur.execute("INSERT INTO foo VALUES (-1)")
 
+    # Wait until WAL is replayed by pageserver
```
Comment (hlinnaka) [inline test_runner/regress/test_lsn_mapping.py:39]:
```suggestion
    # Wait until WAL is received by pageserver
```

File: pgxn/neon/pagestore_smgr.c [hunk 3]
```diff
@@ -1431,7 +1434,11 @@ zenith_truncate(SMgrRelation reln, ForkNumber forknum, BlockNumber nblocks)
 	 */
 	XLogFlush(lsn);
 
-	SetLastWrittenPageLSN(lsn);
+	/*
+	 * Truncate may affect several chunks of relations. So we should either update last written LSN for all of them,
+	 * either update LSN for "dummy" metadata block. Second approach seems to be more efficient.
+	 */
```
Comment (hlinnaka) [inline pgxn/neon/pagestore_smgr.c:1440]:
```suggestion
	/*
	 * Truncate may affect several chunks of relations. So we should either update last written LSN for all of them,
	 * or update LSN for "dummy" metadata block. Second approach seems more efficient. If the relation is extended
	 * again later, the extension will update the last-written LSN for the extended pages, so there's no harm in
	 * leaving behind obsolete entries for the truncated chunks.
	 */
```

**Review by hlinnaka — APPROVED** [review]

**Comment by hlinnaka** [comment]
Can you separate the test changes and the code changes into two commits, and push them separately, please? That way, we'll get "before" and "after" measurements from the CI.

**Comment by knizhnik** [comment]
> Can you separate the test changes and the code changes into two commits, and push them separately, please? That way, we'll get "before" and "after" measurements from the CI.

Do you mean create separate PR for tests or just add tests in separate commit within this PR?
But in the last case I afraid that "squash and marge" will still group them in one commit...

**Comment by hlinnaka** [comment]
> > Can you separate the test changes and the code changes into two commits, and push them separately, please? That way, we'll get "before" and "after" measurements from the CI.
> 
> Do you mean create separate PR for tests or just add tests in separate commit within this PR? But in the last case I afraid that "squash and marge" will still group them in one commit...

You can use squash them into two commits on your laptop with "git rebase -i origin/master", and force push to the branch to update the PR. After the CI passes,  you can do a regular push from your laptop with "git push origin last_written_lsn_cache:main".

Hmm. Not sure the CI would run on 'main' separately for the two commits if you do that though. So two separate PRs is probably the easiest way to do it.

**Comment by knizhnik** [comment]
> > > Can you separate the test changes and the code changes into two commits, and push them separately, please? That way, we'll get "before" and "after" measurements from the CI.
> > 
> > 
> > Do you mean create separate PR for tests or just add tests in separate commit within this PR? But in the last case I afraid that "squash and marge" will still group them in one commit...
> 
> You can use squash them into two commits on your laptop with "git rebase -i origin/master", and force push to the branch to update the PR. After the CI passes, you can do a regular push from your laptop with "git push origin last_written_lsn_cache:main".
> 
> Hmm. Not sure the CI would run on 'main' separately for the two commits if you do that though. So two separate PRs is probably the easiest way to do it.

I have merged PR with tests and rebased this PR.
But to be able to merge it I need your approval for its postgres repository part:
https://github.com/neondatabase/postgres/pull/201

**Pull request merged**